// eslint-plugin-better-tailwindcss/lib/cjs/rules/deprecated/multiline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.multiline = void 0;
const enforce_consistent_line_wrapping_js_1 = require("../enforce-consistent-line-wrapping.js");
exports.multiline = {
    ...enforce_consistent_line_wrapping_js_1.enforceConsistentLineWrapping,
    name: "multiline",
    rule: {
        ...enforce_consistent_line_wrapping_js_1.enforceConsistentLineWrapping.rule,
        meta: {
            ...enforce_consistent_line_wrapping_js_1.enforceConsistentLineWrapping.rule.meta,
            deprecated: {
                availableUntil: "^4.0.0",
                deprecatedSince: "^3.4.0",
                replacedBy: [
                    {
                        message: "The rule name `multiline` is deprecated. Please use `enforce-consistent-line-wrapping` instead.",
                        rule: {
                            name: "enforce-consistent-line-wrapping"
                        }
                    }
                ]
            }
        }
    }
};
